src/all/snowmtl/src/eu/kanade/tachiyomi/extension/all/snowmtl/interceptors/TranslationInterceptor.kt

```diff
@@ -0,0 +1,141 @@
+package eu.kanade.tachiyomi.extension.all.snowmtl.interceptors
+
+import android.os.Build
+import androidx.annotation.RequiresApi
+import eu.kanade.tachiyomi.extension.all.snowmtl.Dialog
+import eu.kanade.tachiyomi.extension.all.snowmtl.Snowmtl.Companion.PAGE_REGEX
+import eu.kanade.tachiyomi.extension.all.snowmtl.Source
+import eu.kanade.tachiyomi.extension.all.snowmtl.translator.TranslatorEngine
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+import okhttp3.Interceptor
+import okhttp3.Response
+import uy.kohesive.injekt.injectLazy
+
+@RequiresApi(Build.VERSION_CODES.O)
+class TranslationInterceptor(
+    private val source: Source,
+    private val translator: TranslatorEngine,
+) : Interceptor {
+
+    private val json: Json by injectLazy()
+
+    override fun intercept(chain: Interceptor.Chain): Response {
+        val request = chain.request()
+        val url = request.url.toString()
+
+        if (PAGE_REGEX.containsMatchIn(url).not() || source.target == source.origin) {
+            return chain.proceed(request)
+        }
+
+        val dialogues = request.url.fragment?.parseAs<List<Dialog>>()
+            ?: return chain.proceed(request)
+
+        val translated = translationOptimized(dialogues)
+
+        val newRequest = request.newBuilder()
+            .url("${url.substringBeforeLast("#")}#${json.encodeToString(translated)}")
+            .build()
+
+        return chain.proceed(newRequest)
+    }
+
+    /**
+     * Optimizes the translation of a list of dialogues.
+     * This reduces the number of requests to the translator per page.
+     *
+     * @param dialogues List of Dialog objects to be translated.
+     * @return List of translated Dialog objects.
+     */
+    private fun translationOptimized(dialogues: List<Dialog>): List<Dialog> {
+        val mapping = buildMap(dialogues)
+
+        val tokens = tokenizeAssociatedDialog(mapping).flatMap { token ->
+            translator.translate(source.origin, source.target, token).split(delimiter)
+        }
+
+        return replaceDialoguesWithTranslations(tokens, mapping)
+    }
```

Comment:
correct me if wrong but what essentially happening is that you merge multiple dialogs or tokens using delimeters as long as it is withing capacity of the translator and send translation request for each of these merged tokens?

Reply:
> correct me if wrong but what essentially happening is that you merge multiple dialogs or tokens [..]

yes, for example, chinese manhuas have many dialogues per page